### ISSUE TITLE
[RFC]: implementation modification to improve accuracy of `@stdlib/math/base/special/binomcoef`

### ISSUE THREAD

**Issue by Andrew-Cottrell** [stdlib-js/stdlib#1155]
### Description

This RFC proposes modifying the implementation of `@stdlib/math/base/special/binomcoef` to improve accuracy.

Example inaccuracies with the [current implementation](https://github.com/stdlib-js/stdlib/blob/d54143e5d29e8c320914df3e989dab7f923a5b31/lib/node_modules/%40stdlib/math/base/special/binomcoef/lib/main.js):
```javascript
binomcoef( 54, 23 ); // expected: 1085929983159840, actual: 1085929983159839
binomcoef( 55, 24 ); // expected: 2488589544741300, actual: 2488589544741299
binomcoef( 56, 22 ); // expected: 2142582442263900, actual: 2142582442263899
binomcoef( 58, 20 ); // expected: 1847253511032930, actual: 1847253511032931
```

I suggest modifying the implementation to use `@stdlib/math/base/special/gcd` in

```javascript
var isInteger = require( '@stdlib/math/base/assert/is-integer' );
var isnan = require( '@stdlib/math/base/assert/is-nan' );
var isOdd = require( '@stdlib/math/base/assert/is-odd' );
var gcd = require( '@stdlib/math/base/special/gcd' );

        ...

	var res;
	var j;
	var d; // 'd' for divisor

        ...

	res = n;
	for ( j = 2; j <= k; j++ ) {
		d = gcd( res, j );
		res = ( res / d ) * ( d * ( n - j + 1 ) / j );
	}
	return res;
```

Notes:
- Exact when the binomial coefficient is not greater than `Number.MAX_SAFE_INTEGER`.
- Reasonably accurate when the binomial coefficient is not greater than `Number.MAX_VALUE`.
- `res` is always an integer or `Infinity`, which removes the need to correct for rounding errors.

Arguments against:
- Accurate large binomial coefficients are unimportant for most (all?) use cases involving computation of binomial coefficients.
- Reduces performance due to repeated calls to `gcd`.

Arguments for:
- Generally, it is preferable for algorithms to be more accurate when possible.
- Performance of computation of binomial coefficients might be unimportant for many use cases (e.g., only a moderate number of binomial coefficients computed, and overall performance dominated by other computations).

### Related Issues

https://github.com/stdlib-js/stdlib/blob/0e370957ae1d899c2d07a67ab92c5d45197c78b6/lib/node_modules/%40stdlib/math/base/tools/fibpoly/test/test.js#L83

https://github.com/stdlib-js/stdlib/blob/0e370957ae1d899c2d07a67ab92c5d45197c78b6/lib/node_modules/%40stdlib/math/base/tools/fibpoly/test/test.factory.js#L89

#802 (search for "binomcoef")

### Questions

No.

### Checklist

- [X] I have read and understood the [Code of Conduct](https://github.com/stdlib-js/stdlib/blob/develop/CODE_OF_CONDUCT.md).
- [X] Searched for existing issues and pull requests.
- [X] The issue name begins with `RFC:`.

**Comment by Andrew-Cottrell** [comment]
An alternative version of the modification avoids calling `gcd` when unnecessary, at the cost of increased code complexity. 

```javascript

	var res;
	var j;
	var s; // 's' for safe
	var d; // 'd' for divisor

        ...

	s = Math.floor(Number.MAX_SAFE_INTEGER / n);
	res = n;
	for ( j = 2; j <= k && res < s; j++ ) {
		res = ( res * ( n - j + 1 ) ) / j;
	}
	for ( ; j <= k; j++ ) {
		d = gcd( res, j );
		res = ( res / d ) * ( d * (n - j + 1) / j );
	}
	return res;
```

The next version swaps the order of the _numerator_ values to be increasing rather than decreasing, which helps performance by sometimes evaluating more terms using the non-`gcd` version of the calculation.

```javascript
	s = Math.floor(Number.MAX_SAFE_INTEGER / n);
	res = n - k + 1;
	n = res + 1;
	for ( j = 2; j <= k && res < s; j++, n++ ) {
		res = (res * n) / j;
	}
	for ( ; j <= k; j++, n++ ) {
		d = gcd( res, j );
		res = ( res / d ) * ( d * n / j );
	}
	return res;
```

In some non-scientific testing, this version appears to be about three times faster, than the simple modification, when computing the four example binomial coefficients mentioned above. And about seven times faster when computing smaller binomial coefficients, such as `binomcoef( 54, 20 )`, that don't need to call the `gcd` function.

**Comment by kgryte** [comment]
@Andrew-Cottrell Thanks for opening this issue and your proposed solutions. Based on my testing, performance drops by over 10x when moving to GCD, even for the improved version.

The accuracy does improve, which is a plus, but I'd prefer, if possible, to find a different algorithm which doesn't force an order of magnitude drop in performance, especially as `binomcoef` is used as a dependency by other functions within `stdlib`. Meaning, a significant drop in `binomcoef` perf will likely necessarily mean a drop in dependants perf.

I've done a bit of research to try and find other algorithms which improve accuracy without sacrificing perf; however, I haven't had much luck finding one which is universally good.

**Comment by kgryte** [comment]
@Andrew-Cottrell I did a bit more investigating. I was able to update the implementation to only compute the GCD when overflow is a risk. For larger binomial coefficients, this drops performance by 50%. Not great, but also not the end of the world.

For reference: https://github.com/stdlib-js/stdlib/blob/9782026baac4239e8fabd217952b77bd633bce75/lib/node_modules/%40stdlib/math/base/special/binomcoef/lib/main.js#L99

Thanks for opening this issue and helping point in the right direction. This resolves the failing tests in `fibpoly`, which have subsequently been updated.
